tests/testsuite/pub_priv.rs

```diff
@@ -198,6 +193,47 @@ Caused by:
         .run()
 }
 
+#[cargo_test]
+fn pub_dev_dependency_without_feature() {
+    Package::new("pub_dep", "0.1.0")
+        .file("src/lib.rs", "pub struct FromPub;")
+        .publish();
+
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+                [package]
+                name = "foo"
+                version = "0.0.1"
+
+                [dev-dependencies]
+                pub_dep = {version = "0.1.0", public = true}
+            "#,
+        )
+        .file(
+            "src/lib.rs",
+            "
+            extern crate pub_dep;
+            pub fn use_pub(_: pub_dep::FromPub) {}
+        ",
+        )
+        .build();
+
+    p.cargo("check --message-format=short")
+        .masquerade_as_nightly_cargo(&["public-dependency"])
+        .with_status(101)
+        .with_stderr(
+            "\
+[ERROR] failed to parse manifest at `[..]`
+
+Caused by:
+  'public' specifier can only be used on regular dependencies, not Development dependencies
+",
+        )
+        .run()
+}
```

Comment:
Now that we have a test to more concretely talk about this, I'm going to continue on [this comment thread](https://github.com/rust-lang/cargo/pull/13340#discussion_r1463618205) here.

Reply:
> ... we should probably check the unstable status and report this as a warning that this will fail when stabilized. Or in other words, without the -Z, users should get the experience they have before any work was done on this but be warned what will break

Our options are
1. Validate and error like we do here
2. Validate but warn on stable
3. Do nothing, relying only on the "this isn't stable / unused" warning

This takes (1).  I had thought `[lints]` went with (2) but it in looking at [the code](https://github.com/rust-lang/cargo/blob/0.74.0/src/cargo/util/toml/mod.rs#L2833), it went with (3).  Unsure why I changed my mind.

We should make a conscious choice about what route we go.
- The downside to errors (1) is that changes in unstable behavior affect users on stable.
- Warning (2) can help people on mixed stable/nightly detect changes earlier
- For stable-only people with an MSRV, the warning (2) will be unactionable and annoying

Reply:
`that changes in unstable behavior affect users on stable.`  
Not quite sure how this happened, and could it have been avoided?

Reply:
> So those tests show that we are erroring on this unstable feature when we should instead warn that this will error out in the future when its stabilized.

```
  if dep.kind() != DepKind::Normal {
            bail!("'public' specifier can only be used on regular dependencies, not {:?} dependencies", dep.kind());
```
Change this to a warning?

Reply:
> Not quite sure how this happened, and could it have been avoided?

If someone is using an unstable feature on a specific nightly but wants it to work generally on stable, we need to be careful.  However, that is more about introducing new errors.  Removing an error (the main risk here) wouldn't be breaking.

> Change this to a warning?

When the feature isn't enabled?  That is option (2)

Reply:
I had made a new commit.
if a dependency isn't normal dependency,  warn `public` used in dev-dependency when no feature enabled in nightly.   we should to skip set `dep.set_public(p)`, and set `orig.public = None`.